### ISSUE TITLE
Extra '/' at the end of the request URL

### ISSUE THREAD

**Issue by Nechitadi** [jnunemaker/httparty#742]
Hello,

I've noticed that starting from version v0.18.0 there is an extra '/' added at the end of the request url. Is this something on purpose? Maybe it works for the REST api, but we are using SOAP api on our project and it does not accept any '/' at the end.

**Comment by jnunemaker** [comment]
@Nechitadi I can't replicate this. Can you get me a failing script or spec? If you change the uri adapter to addressable does it work? 

```ruby
uri_adapter(Addressable::URI)
```

**Comment by Nechitadi** [comment]
Yes, it worked with the uri_adapter(Addressable::URI). Is there any way to set this option by default? Because we have multiple services that use HTTParty for soap. Or just by creating a base service and the other ones to inherit from there?

**Comment by jnunemaker** [comment]
Yep, you’ll need a base service or another include that sets the uri
adapter. 

Must be Ruby’s uri has an issue with this as that is the default uri
adapter. 

On November 10, 2021, David Grevink ***@***.***> wrote:
> Yes, it worked with the uri_adapter(Addressable::URI). Is there any
> way to set this option by default? Because we have multiple services
> that use HTTParty for soap. Or just by creating a base service and the
> other ones to inherit from there?
>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/jnunemaker/httparty/issues/742#issuecomment-
964849688>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AAAAB27WJ5R6NKVO6CLEYZDULIMKDANCNFSM5HUXXTDQ>.

**Comment by ozgun** [comment]
Hello, 

We experience the same issue.

When we set the base URI with `base_uri('http://example.com/api/articles')` and make a `GET` request with the `get` method by passing an empty string(`''`) as the `path`, then the URL is appended a forward slash and is converted to `http://example.com/api/articles/`.

```ruby
get("")
```

As far as I traced down, the following change, which was introduced with [this commit](https://github.com/jnunemaker/httparty/commit/54b7949b91b045680ea4cd40a91d2bad7165a295#diff-ffdb28e32ba3c866342867c542bb2b917c601cffdfe3bb5f1321147a62c6a0e1) in version 0.17.1, is causing the issue. The following line in the `path=` method converts the empty string to `/` after the `normalize` method is called:

```ruby
uri_adapter.parse(uri).normalize
```

As @jnunemaker suggested, changing the uri adapter to `Addressable::URI` seems to be fixing the issue:

```ruby
Addressable::URI.parse("").normalize.path #=> ""

URI.parse("").normalize.path  #=> "/"
```

However, I wonder if we could solve this issue by skipping the normalization when the uri is an empty string by updating the `path=` method in `lib/httparty/request.rb` file as below:

```ruby
def path=(uri)
  uri_adapter = options[:uri_adapter]

  @path = if uri.is_a?(uri_adapter)
    uri
  elsif String.try_convert(uri)
    if uri == ""
      uri_adapter.parse(uri)
    else
      uri_adapter.parse(uri).normalize
    end
  else
    raise ArgumentError,
      "bad argument (expected #{uri_adapter} object or URI string)"
  end
end
```

If you think this is a valid solution, I can create a PR, thanks!

**Comment by jnunemaker** [comment]
@ozgun Should we just move the normalize to the uri adapter? Then people could use whatever uri_adapter they want to normalize or not normalize. Thoughts? Yes, I'm open to PR to see the full scope of spec and code changes. :)
